test/stdlib/TestJSONEncoder.swift

```diff
@@ -231,11 +230,65 @@ class TestJSONEncoder : TestJSONEncoderSuper {
 
   // MARK: - Date Strategy Tests
   func testEncodingDate() {
+
+    func formattedLength(of value: Double) -> Int {
+      let empty = UnsafeMutablePointer<Int8>.allocate(capacity: 0)
+      defer { empty.deallocate() }
+      let length = snprintf(ptr: empty, 0, "%0.*g", DBL_DECIMAL_DIG, value)
+      return Int(length)
+    }
+
+    // Duplicated to handle a special case
+    func localTestRoundTrip<T: Codable>(of value: T) {
+      var payload: Data! = nil
+      do {
+        let encoder = JSONEncoder()
+        payload = try encoder.encode(value)
+      } catch {
+        expectUnreachable("Failed to encode \(T.self) to JSON: \(error)")
+      }
+
+      do {
+        let decoder = JSONDecoder()
+        let decoded = try decoder.decode(T.self, from: payload)
+
+        /// `snprintf`'s `%g`, which `JSONSerialization` uses internally for double values, does not respect 
+        /// our precision requests in every case. This bug effects Darwin, FreeBSD, and Linux currently
+        /// causing this test (which uses the current time) to fail occasionally.
+        let evalEdgeCase: (Date, Date) -> () = { decodedDate, expectedDate in
+          if formattedLength(of: decodedDate.timeIntervalSinceReferenceDate) > DBL_DECIMAL_DIG + 2 {
+            let adjustedTimeIntervalSinceReferenceDate: (Date) -> Double = {
+                let adjustment = pow(10, Double(DBL_DECIMAL_DIG))
+                return Double(floor(adjustment * $0.timeIntervalSinceReferenceDate) / adjustment)
+            }
+            
+            let decodedAprox = adjustedTimeIntervalSinceReferenceDate(decodedDate)
+            let valueAprox = adjustedTimeIntervalSinceReferenceDate(expectedDate)
+            expectEqual(decodedAprox, valueAprox, "\(T.self) did not round-trip to an equal value after DBL_DECIMAL_DIG adjustment \(decodedAprox) != \(valueAprox).")
+          }
+        }
+
+        if let decodedDate = (decoded as? TopLevelWrapper<Date>)?.value, 
+           let expectedDate = (value as? TopLevelWrapper<Date>)?.value {
+          evalEdgeCase(decodedDate, expectedDate)
+        }
+
+        if let decodedDate = (decoded as? OptionalTopLevelWrapper<Date>)?.value, 
+           let expectedDate = (value as? OptionalTopLevelWrapper<Date>)?.value {
+          evalEdgeCase(decodedDate, expectedDate)
+        }
+
+        expectEqual(decoded, value, "Date did not round-trip to an equal value.")
```

Comment:
Yeah this will need a tweak. Will post this shortly.

Reply:
Moving this from the playground I tried this on I forgot the `Equatable` before I pushed this up. Addressed.